File: 207 (Course Schedule)/cpp/Solution.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj_list(numCourses);
        vector<int> degree(numCourses,0);
        for(auto val: prerequisites){
            adj_list[val[1]].push_back(val[0]);
            degree[val[0]]++;
        }

        queue<int> validCourses;
        unordered_set<int> isVisited;
        for(int i =0;i<numCourses;i++){
            if (degree[i]==0) {
                validCourses.push(i);
                isVisited.insert(i);
            }
        }

        while(!validCourses.empty()){
            int data = validCourses.front();
            validCourses.pop();
            for(int child : adj_list[data]){
                degree[child]--;
                if (degree[child]==0) {
                    validCourses.push(child);
                    isVisited.insert(child);
                }
            }
        }

        return isVisited.size()==numCourses;
    }
};

File: 207 (Course Schedule)/python/Solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj_list = [[] for _ in range(numCourses)]
        degree = [0 for _ in range(numCourses)]

        for c1,c2 in prerequisites:
            adj_list[c2].append(c1)
            degree[c1]+=1
        
        validCourses = [val for val in range(numCourses) if degree[val]==0]

        isVisited = set(validCourses)

        while validCourses:
            course = validCourses.pop()

            for child in adj_list[course]:
                degree[child]-=1
                if degree[child]==0:
                    isVisited.add(child)
                    validCourses.append(child)

        return len(isVisited)==numCourses